10-H_Device-initiated_Communication_with_NVSHMEM/jacobi.cu

```diff
@@ -0,0 +1,617 @@
+                    "ERROR: CUDA RT call \"%s\" in line %d of file %s failed "              \
+                    "with "                                                                 \
+                    "%s (%d).\n",                                                           \
+                    #call, __LINE__, __FILE__, cudaGetErrorString(cudaStatus), cudaStatus); \
+    }
+
+#ifdef USE_DOUBLE
+typedef double real;
+#define MPI_REAL_TYPE MPI_DOUBLE
+#else
+typedef float real;
+#define MPI_REAL_TYPE MPI_FLOAT
+#endif
+
+constexpr real tol = 1.0e-8;
+
+const real PI = 2.0 * std::asin(1.0);
+
+__global__ void initialize_boundaries(real* __restrict__ const a_new, real* __restrict__ const a,
+                                      const real pi, const int offset, const int nx,
+                                      const int my_ny, const int ny) {
+    for (int iy = blockIdx.x * blockDim.x + threadIdx.x; iy < my_ny; iy += blockDim.x * gridDim.x) {
+        const real y0 = sin(2.0 * pi * (offset + iy) / (ny - 1));
+        a[iy * nx + 0] = y0;
+        a[iy * nx + (nx - 1)] = y0;
+        a_new[iy * nx + 0] = y0;
+        a_new[iy * nx + (nx - 1)] = y0;
+    }
+}
+
+template <int BLOCK_DIM_X, int BLOCK_DIM_Y>
+__global__ void jacobi_kernel(real* __restrict__ const a_new, const real* __restrict__ const a,
+                              real* __restrict__ const l2_norm, const int iy_start,
+#ifdef SOLUTION
+                              const int iy_end, const int nx, const bool calculate_norm,
+                              const int top, const int iy_top_lower_boundary_idx,
+                              const int bottom, const int iy_bottom_upper_boundary_idx) {
+#else
+                              const int iy_end, const int nx, const bool calculate_norm) {
+#endif
+#ifdef HAVE_CUB
+    typedef cub::BlockReduce<real, BLOCK_DIM_X, cub::BLOCK_REDUCE_WARP_REDUCTIONS, BLOCK_DIM_Y>
+        BlockReduce;
+    __shared__ typename BlockReduce::TempStorage temp_storage;
+#endif  // HAVE_CUB
+    int iy = blockIdx.y * blockDim.y + threadIdx.y + iy_start;
+    int ix = blockIdx.x * blockDim.x + threadIdx.x + 1;
+    real local_l2_norm = 0.0;
+
+    if (iy < iy_end && ix < (nx - 1)) {
+        const real new_val = 0.25 * (a[iy * nx + ix + 1] + a[iy * nx + ix - 1] +
+                                     a[(iy + 1) * nx + ix] + a[(iy - 1) * nx + ix]);
+        a_new[iy * nx + ix] = new_val;
+        if (calculate_norm) {
+            real residue = new_val - a[iy * nx + ix];
+            local_l2_norm += residue * residue;
+        }
+
+#ifdef SOLUTION
+        if (iy_start == iy) {
+            nvshmem_float_p(a_new + iy_top_lower_boundary_idx * nx + ix, new_val, top);
+        }
+        if ((iy_end - 1) == iy) {
+            nvshmem_float_p(a_new + iy_bottom_upper_boundary_idx * nx + ix, new_val, bottom);
+        }
+#else
+        //TODO: push values near boundary to top and bottom PE, remember to change the signature of
```

Comment:
I like putting the `//TODO:` parts outside the `#ifdef SOLUTION` so I have an easier time finding the correct position when presenting solutions in class. Just an FYI.
Let me know if I should merge like it is or if you want to change that.